### ISSUE TITLE
Pseudopotential (-numerical atomic orbital) test database will be available after ABACUS 3.5.0

### ISSUE THREAD

**Issue by kirk0830** [deepmodeling/abacus-develop#3159]
### Details

Materials Cloud/Standard solid-state pseudopotentials (SSSP) is a good example for user to quickly check availabilities, correctness and accurancy estimation of pseudopotentials that are supported by Quantum ESPRESSO. User can also find ecutwfc value on it so that convergence test on this parameter is not needed for not-too-strict calculations.
We will also have such a database open for ABACUS users for parameters such as `ecutwfc`, and choice of different kinds of pseudopotentials and corresonding numerical orbitals. Numerical orbitals would be tested from the lowest level DZP to TZDP, the shortest cutoff radius 6 a.u. to the longest 10 a.u. presently.

I am a little bit curious about systems to test and physical properties to calculate. Of course band gap and stress are now considered as indices to test, but what about others?
The case to test should not be too specific, although specific case can be tested on the actual user need, and will be collected.

Any ideas will be appreciated.

### Have you read the online manual http://abacus.deepmodeling.com/en/latest/

- [X] Yes, I have read the online manual.

### Task list for Issue attackers (only for developers)

- [X] Identify the specific section of the documentation with the issue.
- [X] Investigate the issue and determine the root cause.
- [X] Research best practices and potential solutions for the identified issue.
- [X] Update the documentation to address the issue, following the suggested improvement.
- [ ] Ensure the updated documentation adheres to the project's documentation standards.
- [ ] Test the updated documentation to ensure it is clear and accurate.
- [ ] Review and incorporate any relevant feedback from users or developers.
- [ ] Publish the updated documentation and notify the issue reporter.

**Comment by kirk0830** [comment]
@QG-phys provides SC/40 semiconductor test set:
Heyd, J., Peralta, J. E., Scuseria, G. E., & Martin, R. L. (2005). Energy band gaps and lattice parameters evaluated with the Heyd-Scuseria-Ernzerhof screened hybrid functional. The Journal of Chemical Physics, 123(17), 174101. doi:10.1063/1.2085170

**Comment by kirk0830** [comment]
@QuantumMisaka Do you have any ideas? I think any training set of machine learning training set will be a not bad choice. Such that those labels to trained for will also be good choices to show convergence on some variables

**Comment by QuantumMisaka** [comment]
In Catalysis system, these type of catalysts, and adsorption or reaction on them is typical:
- simple surface model, like Pt(111), MgO(110)
- large metal-oxides or other framework-like system, including ZSM-5 or MOF
- Metal-doped graphene system
- systems with huge magnetic moment like Fe5C2(510) in FTS process
- systems with strong-correlated effect like CeO2 surface

Typical properties for concern:
- Most important: Optimized Structure (usually have large number of atoms and vacuum site) and Total Energy (also Forces, sometimes Stress)  for surface, adsorption state and transition state. 
- Free Energy from correction of calculated total energy (which can be done by `Phonopy`)
- Electronic structure information is sometimes concerned, like Charge (Bader is commonly used), spin-sum-ed Charge, spin-split-ed Charge, DOS (PDOS or LDOS) and COHP. but the Energy Band is less concerned.
- Magnetic Moment number for atoms of magnetic system

Training set of ML is not a bad choice but there will be much more redundant information in there. may be some typical system from published paper can be more effective ?

I'm preparing some examples from catalysis system, which can be upload if needed

**Comment by kirk0830** [comment]
A small demo:

# ABACUS psesudopotentials and numerical orbitals test workflow
*Author: kirk0830*
## Environments and dependencies
### Executables
* Quantum ESPRESSO `I1d.x` executable is needed to generate PSlibrary 0.3.1 and 1.0.0 version PAW and US pseudopotential files. The executable is not included in the repository.
* PTG_dpsi executable is needed to generate numerical orbitals for specific kinds of pseudopotential.
* ABACUS `abacus` executable is needed to generate numerical orbitals that provides reference wavefunction.
### Python packages
* `numpy` is needed for handling numerical data.
* `scipy` is needed for some special functions.
* `matplotlib` is needed for plotting.
* `json` and `time` are needed for storing work logs.
### Python program list
* `pseudopot_test_generation.py`: generate pseudopotential files for specific kinds of pseudopotential, use Quantum ESPRESSO HSE functional.
* `nao_test_generation.py`: generate numerical orbitals for specific kinds of pseudopotential. Use ABACUS.
* `pseudopot_test_analysis.py`: band structure analysis, calculates band gap and plot
* `pseudopot_test_fold_band.py`: band folding, calculates DOS
### File structure
```
├── readme.md
├── resources.json
├── pseudopot_test_generation.py
├── nao_test_generation.py
├── pseudopot_test_analysis.py
├── pseudopot_test_fold_band.py
├── numerical_orbitals/
│  ├── resources/
│  │   ├── sg15_10 ([pseudopot_kind]_[pseudopot_version]_[appendix])
│  │   │   ├── 33_As_DZP ([atomic_number]_[element]_[basis_set])
│  │   │   │   ├── As_gga_7au_100Ry_2s2p1d.orb
│  │   │   │   └── ...
│  │   │   └── ...
│  │   ├── ...
│  │   └── tools/
│  │       ├── atom_in.py
│  │       ├── orb_io.py
│  │       ├── cal_ovlp_jjtilde.py
│  │       ├── ABACUS_orb_parser.py
│  │       ├── nao_oscillation_remote.py
│  │       └── ...
│  ├── results/
│  │   ├── ...
│  │   └── ...
│  ├── templates/
│  │   ├── INPUT
│  │   ├── KPT
│  │   └── STRU_* (for each system to test)
│  ├── config.json (for ABACUS Test setting)
│  └── ...
├── pseudopotentials/
│  ├── resources/
│  │   ├── sg15_10 ([pseudopot_kind]_[pseudopot_version]_[appendix])
│  │   ├── pd_03
│  │   │   ├── As_*.upf
│  │   │   └── ...
│  │   ├── pd_04
│  │   ├── pd_04_d
│  │   └── ...
│  ├── results/
│  │   ├── t_*
│  │   └── ...
│  ├── templates/
│  │   └── template_*.in (for each system to test)
│  ├── config.json (for ABACUS Test setting)
│  └── ...
├── analysis/
│  └── [developing]
└── ...
```
## Usage
### Configuration
#### Files auto-download
In `./resources.json`, add/change `resources` section to download files automatically. `jsons` section will be useful when preparing input scripts.
#### Pseudopotential generation
For cases such as `kjpaw`, `rrkjus` or `pslnc` those pseudopotential kinds belonging to PSlibrary, pseudopotential may need to generate with Quantum ESPRESSO `Id1.x` code. Therefore this executable must be ready. If so, change the variable `qe_path` in `resources.json` to the path of folder, where there is `Id1.x` executable.
#### Numerical orbital generation
For cases such as non-SG15, numerical atomic orbital is needed to generate. Please follow the handson of PTG_dpsi to install executables of ABACUS, PTG_dpsi and other dependencies. Change the variable `abacus_path` in `resources.json` to the path of folder, where there is `abacus` executable.  
Numerical atomic orbitals of other kinds of pseudopotentials will be generated based on parameters set in `SIAB_INPUT` of SG15-kind pseudopotential corresponding numerical atomic orbital generation.
### Input scripts generation
* execute `pseudopot_test_generation.py` to generate test input scripts.
* execute `nao_test_generation.py` to generate numerical orbitals.
### Test run
visit ABACUS Test website developed by ABACUS team to submit test jobs.  
website: https://labs.dp.tech/projects/abacustest/
#### Pseudopotential test
* use `normal| rundft, postdft` module, upload a zip file. For more test configurations, please refer to `./pseudpotentials/config.json`.
#### Numerical orbital test
* use `expert| predft, rundft, postdft` module, upload a zip file. For more test configurations, please refer to `./numerical_orbitals/config.json`.
### Analysis
* execute `pseudopot_test_analysis.py` to analyze band structure and calculate band gap. For more functions of it, please read in-file comments.
## Phylosophy of organizing information in jsons
### resources.json
TBD...
### config.json
TBD...
### work_status
`work_status` is the input variable of `pseudopot_test_generation.py`, `pseudopot_test_analysis.py` and `nao_test_generation.py`. It always has such following keywords:
```
{
    "work_folder": "path/to/work/folder",
    "pseudopot_kinds": ["pseudopot_kind_1", "pseudopot_kind_2", ...],
    "versions": ["version_1", "version_2", ...],
    "appendices": ["appendix_1", "appendix_2", ...],
    ...
    "systems": {
        "system1": {
            "experimental_value": 1.0,
        }
        "system2": {
            "experimental_value": 2.0,
        }
        ...
    }
}
```
In this json, user should give enough information for the whole test, included but not limited to systems description, sometimes the postprocessing such as plot is also needed.
### test_status
test_status is a on-the-fly or in-build json, which will pass from one function to another inside program. If there is no special case, it will not be a real output. test_status describes with more details not only for systems, but may also carry data.
The design of test_status is shown in following:
```
{
    "work_folder": "./numerical_orbitals",
    "InAs": {
        "sg15_10_7_DZP_pbe": {
            "elements": [
                "In",
                "As"
            ],
            "pseudopotentials": [
                "In_ONCV_PBE-1.0.upf",
                "As_ONCV_PBE-1.0.upf"
            ],
            "numerical_orbitals": [
                "In_gga_7au_100Ry_2s2p2d1f.orb",
                "As_gga_7au_100Ry_2s2p1d.orb"
            ]
        }
}
```
Therefore except `work_folder` and `root`, other keys are all system names. Under each system, there are test names, and under each pseudopotential, there are elements and numerical orbitals.  
- For pseudopotential tests, the name of test is organized as [pseudopot_kind]_ [pseudopot_version]_ [appendix]_ [functional].  
- For numerical atomic orbital tests, the name of test is organized as [pseudopot_kind]_ [pseudopot_version]_ [appendix]_ [cutoff radius]_ [zeta information]_ [functional].
## Other tools
### Numerical atomic orbital analysis
Due to present method to generate numerical atomic orbitals has some problems, it is necessary to check the quality of generated numerical atomic orbitals. `./numerical_orbitals/resources/tools/ABACUS_orb_parser.py` is a tool to check the oscillation of numerical atomic orbitals.  
If indeed there is high frequency, unphysical oscillation in orbitals generated, use `./numerical_orbitals/resources/tools/nao_oscillation_remote.py` to remove high frequency oscillation and generate new numerical atomic orbitals.

**Comment by kirk0830** [comment]
It is avaiable here, but not complete yet: https://github.com/kirk0830/ABACUS-pseudopotential-square

**Comment by kirk0830** [comment]
issue https://github.com/deepmodeling/abacus-develop/issues/2243 mentioned precision autotest workflow will be a part of APNS.

**Comment by kirk0830** [comment]
Create three workflows in APNS:
![image](https://github.com/deepmodeling/abacus-develop/assets/67682086/7b01e044-ea64-4c22-9460-22e4ae03b5a5)
, supported by Python polymorphism:
```Python
"""APNS has three main functionalities: test, analysis and orbital generation. Each functionality has its own driver."""

class apns_driver:
    """APNS driver abstract class, unifying workflow drivers.
    For each kind of driver, must setup first, then run."""
    def __init__(self, finp: str):
        self.finp = finp
    def setup(self):
        """setup the driver"""
        pass
    def run(self):
        """run the driver"""
        pass

import apns.module_workflow.workflow_test.driver as amwtd
class test_driver(apns_driver):
    """test driver, for testing pseudopotentials and numerical orbitals"""
    def setup(self):
        """setup the driver"""
        pass
    def run(self):
        amwtd.driver_v1(self.finp)

import apns.module_workflow.workflow_analysis.driver as amwad
class analysis_driver(apns_driver):
    """analysis driver, for analyzing test results"""
    def setup(self):
        """setup the driver"""
        pass
    def run(self):
        pass

import apns.module_workflow.workflow_orbgen.driver as amwod
class orbgen_driver(apns_driver):
    """orbgen driver, for generating numerical orbitals"""
    def setup(self):
        """setup the driver"""
        pass
    def run(self):
        pass

import json
def spawn_driver(finp: str) -> apns_driver:
    """return corresponding driver according to detailed user settings"""
    with open(finp, "r") as f:
        inp = json.load(f)
    if inp["global"]["test_mode"] == "pseudopotential" or inp["global"]["test_mode"] == "numerical_orbital":
        print("Activate test mode: ", inp["global"]["test_mode"])
        return test_driver(finp)
    elif inp["global"]["test_mode"] == "analysis":
        print("Analysis mode activated.")
        return analysis_driver(finp)
    elif inp["global"]["test_mode"] == "orbgen":
        print("Orbgen mode activated.")
        return orbgen_driver(finp)
    else:
        raise ValueError("Invalid test mode.")
```

**Comment by kirk0830** [comment]
APNS has been published on 2024/04/01, further updates will be posted on new issues.
